File: 백준/Gold/1967. 트리의 지름/트리의 지름.java
import java.io.*;
import java.util.*;

public class Main {
    static List<Node>[] tree;
    static boolean[] visited;
    static int maxDist = 0;
    static int farthestNode = 0;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        if (n == 1) {
            System.out.println(0);
            return;
        }

        tree = new ArrayList[n + 1]; // 2차원 배열
        for (int i = 0; i <= n; i++) {
            tree[i] = new ArrayList<>();
        }

        for (int i = 0; i < n - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());
            tree[a].add(new Node(b, cost));
            tree[b].add(new Node(a, cost));
        }

        // 1번 노드에서 가장 먼 노드 찾기
        visited = new boolean[n + 1];
        visited[1] = true;
        dfs(1, 0);

        // 찾은 노드에서 가장 먼 노드까지 거리 찾기
        visited = new boolean[n + 1];
        maxDist = 0; // 전역 변수이므로 다시 초기화
        visited[farthestNode] = true;
        dfs(farthestNode, 0);

        System.out.println(maxDist);
    }

    static void dfs(int node, int dist) {
        if (dist > maxDist) {
            maxDist = dist;
            farthestNode = node;
        }

        for (Node next : tree[node]) {
            if (!visited[next.to]) {
                visited[next.to] = true;
                dfs(next.to, dist + next.cost);
            }
        }
    }

    static class Node {
        int to, cost;

        Node(int to, int cost) {
            this.to = to;
            this.cost = cost;
        }
    }
}

File: 백준/Gold/1967. 트리의 지름/트리의 지름.py
import sys
from sys import stdin
sys.setrecursionlimit(10**6)

# 입력 처리
n = int(stdin.readline()) # 노드의 개수

# 노드가 1개일 땐 트리의 지름이 0
if n == 1:
    print(0)
    exit()

tree  = [[] for _ in range(n+1)]

for _ in range(n-1):
    a, b, cost = map(int, stdin.readline().split())
    tree[a].append((b, cost))
    tree[b].append((a, cost))

# 가장 먼 노드 찾기
def dfs(node, dist):
    global max_dist, farthest_node
    
    if dist > max_dist:
        max_dist = dist
        farthest_node = node
    
    # 현재 노드와 연결된 모든 노드 탐색
    for next_node, cost in tree[node]:
        if not visited[next_node]:
            visited[next_node] = True
            dfs(next_node, dist + cost) # 현재까지 이동한 거리(dist)에다가 현재 노드에서 다음 노드로 가는 비용(cost)을 더함

# 1번 노드에서 가장 먼 노드 찾기
visited = [False] * (n + 1)
max_dist, farthest_node = 0, 0
visited[1] = True
dfs(1, 0)

# 1번 노드와 가장 먼 노드에서 가장 먼 노드 찾기
visited = [False] * (n + 1)
max_dist = 0
visited[farthest_node] = True
dfs(farthest_node, 0)

print(max_dist)